orchestrate/obase/pipelines.go

```diff
@@ -0,0 +1,62 @@
+package obase
+
+import (
+	"github.com/relex/gotils/logger"
+	"github.com/relex/gotils/promexporter/promreg"
+	"github.com/relex/slog-agent/base"
+	"github.com/relex/slog-agent/base/bconfig"
+	"github.com/relex/slog-agent/base/bsupport"
+)
+
+// PipelineStarter represents a function to launch workers for a top-level pipeline under Orchestrator
+//
+// bufferID must be unique inside the parent orchestrator
+//
+// Launched workers should start shutting down as soon as the input channel is closed and call onStopped at the end
+type PipelineStarter func(parentLogger logger.Logger, metricCreator promreg.MetricCreator,
+	input <-chan []*base.LogRecord, bufferID string, outputTag string, onStopped func())
+
+// PrepareSequentialPipeline makes a starter for pipelines including transformer, serializer and output forwarder
+func PrepareSequentialPipeline(args bconfig.PipelineArgs) PipelineStarter {
+
+	return func(parentLogger logger.Logger, metricCreator promreg.MetricCreator,
+		input <-chan []*base.LogRecord, bufferID string, outputTag string, onStopped func()) {
+
+		// bufferer in the middle of pipeline has to be started first and shut down last for persistance of pending outputs
+		outputBufferer := args.BufferConfig.NewBufferer(parentLogger, bufferID, args.OutputConfig.MatchChunkID,
+			metricCreator, args.SendAllAtEnd)
+		outputBufferer.Start()
+
+		// then start output forwarder which is at the end of pipeline.
+		// if there are queued logs from bufferer, the consumer would immediately start sending them.
+		if args.NewConsumerOverride != nil {
+			parentLogger.Info("launch override consumer")
+			outputConsumer := args.NewConsumerOverride(parentLogger, outputBufferer.RegisterNewConsumer())
+			outputConsumer.Start()
+		} else {
+			parentLogger.Info("launch consumer")
+			outputConsumer := args.OutputConfig.NewForwarder(parentLogger, outputBufferer.RegisterNewConsumer(), metricCreator)
+			outputConsumer.Start()
+		}
+
+		// then prepare processing worker which is at the head of pipeline
+		procTracker := base.NewLogProcessCounter(metricCreator, args.Schema, args.MetricKeyLocators)
+		procWorker := bsupport.NewLogProcessingWorker(
+			parentLogger,
+			input,
+			args.Deallocator,
+			procTracker,
+			bsupport.NewTransformsFromConfig(args.TransformConfigs, args.Schema, parentLogger, procTracker),
+			args.OutputConfig.NewSerializer(parentLogger, args.Schema, args.Deallocator),
+			args.OutputConfig.NewChunkMaker(parentLogger, outputTag),
+			outputBufferer.Accept,
+		)
+		procWorker.Stopped().Next(func() {
+			outputBufferer.Destroy()
+			onStopped()
+		})
+
+		// only start the processing worker after all the rest have been started
+		procWorker.Start()
```

Comment:
Don't the other workers start by `go buf.feeder.Run()` or simiar? How do you know that they have actually started once execution reaches this point?

Reply:
We don't. That's also related to `IntermediateChannelTimeout` and `IntermediateBufferedChannelSize`, which need to be handle the initial burst of incoming logs.

Normally we should wait and make sure everything is started in the same sequence every time, but any delay of initialization in log agent would likely pause client applications, most of which don't use queues or async operations for logging by default.

To reduce the pause duration, everything are launched almost at the same time and the starting order is only meaningful to protect a few key areas, such as pushing queued logs on disk into output channel before new ones arrive.

Reply:
comment fixed